Normalize comparisons and improve string_t operations

carlopi opened the pull request:
This should only have small positive effects on 2 areas:

- 2-3% binary size reduction (varies on included extensions) on duckdb binary
This is thanks to folding templates into equivalent operators, thanks to the fact that:
```
a >  b --> a > b
a <  b --> b > a
a >= b --> a > b, negated
a <= b --> b > a, negated
```
Changing order of operations is sort of possible in all cases, negating the comparison might need some rework of the calling code like
```
count(a>=b) --> size - count(a>b)
```
that has been partially done, but probably with some effort can be done in even more places.
Secondary advantage is in code reduction.
Test are all passing but one tied to interval >= logic, where given that `==` is different than `nor > nor <', where we move from `>= is > or ==` to a behaviour compliant with postgres. (test is [test/sql/types/interval/test_interval_comparison.test](https://github.com/duckdb/duckdb/compare/feature...carlopi:normalizeComparisons1?expand=1#diff-1749fa6a9596d67becfebc100121b5425d04410f512633f65d4a027408624e16)).

I used [bloaty](https://github.com/google/bloaty), a binary size profiler, to detect relevant highly duplicated templates. Next up are functions like those (and their permutations with other types) but that requires swapping also output-parameters.
```
   0.1%  21.6Ki   0.1%  21.6Ki    unsigned long long duckdb::DistinctSelectGeneric<duckdb::string_t, duckdb::string_t, duckdb::DistinctGreaterThanNullsFirst>(duckdb::Vector&, duckdb::Vector&, duckdb::SelectionVector const*, unsigned long long, duckdb::SelectionVector*, duckdb::SelectionVector*)
   0.1%  21.6Ki   0.1%  21.6Ki    unsigned long long duckdb::DistinctSelectGeneric<duckdb::string_t, duckdb::string_t, duckdb::DistinctGreaterThan>(duckdb::Vector&, duckdb::Vector&, duckdb::SelectionVector const*, unsigned long long, duckdb::SelectionVector*, duckdb::SelectionVector*)
   0.1%  21.6Ki   0.1%  21.6Ki    unsigned long long duckdb::DistinctSelectGeneric<duckdb::string_t, duckdb::string_t, duckdb::DistinctGreaterThanEquals>(duckdb::Vector&, duckdb::Vector&, duckdb::SelectionVector const*, unsigned long long, duckdb::SelectionVector*, duckdb::SelectionVector*)
   0.1%  21.6Ki   0.1%  21.6Ki    unsigned long long duckdb::DistinctSelectGeneric<duckdb::string_t, duckdb::string_t, duckdb::DistinctLessThanNullsFirst>(duckdb::Vector&, duckdb::Vector&, duckdb::SelectionVector const*, unsigned long long, duckdb::SelectionVector*, duckdb::SelectionVector*)
   ```

- perf improvements in string_t related equality, greater-than or hashing functions. In string heavy workloads this should be worth up to 10% improvement (handwavy).
Here we using the earlier simplification to deal with simpler code, and we use the fact that string_t have always 12 bytes of valid access to speed up greater than operator and hashing, and using more heavily the fact that binary equal `string_t`s have to be equal (and the converse it's also true in limited cases, like prefix being different).

Here some hashing-related tests have to be modified for allowing faster computation (expecially on inlined strings where we avoid function calls / and pay only 1 branch to detect inlining).

I believe (should add a test) that in this version hash(string_t) should be the same of the equivalent hash(char*).

- added passing by `const &` in the comparison_operator interface
This should be negligible, but I believe it's cleaner and might, in some rare circumstances, enable better code generation.

carlopi committed:
Move specialize equality and greaterThan opeartor to string_t header

carlopi committed:
string_t equality: check lenght AND prefix with single uint64_t comparison

carlopi committed:
string_t equality: blind check on pointer / inlined lenght fast-path

carlopi committed:
string_t equality: IsInlined will always be different (if gets here), treat !IsInlined

carlopi committed:
Normalization disequalities into equalities

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>'

carlopi committed:
string_t's EqualsOrNot<> do not need template anymore

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>' /2

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>' /3

carlopi committed:
string_t's disequality operator do not need to be templated any more

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>' /4

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>' /5

carlopi committed:
Normalize disequalities: >,>=,< and <= can all be expressed as transformations on '>' /6

carlopi committed:
string_t disequality: Fast path 4-bytes check on prefix

carlopi committed:
string_t disequality: GreaterThan to be nested in StringComparisons (NFC)

carlopi committed:
comparison_operations: Avoid duplicate logic, only Equals and GreaterThan to be implemented

carlopi committed:
string_t (dis)equality: Implement fall-back for NO_INLINE

carlopi committed:
comparisons: Remove NotDistictFrom implementations

carlopi committed:
comparison_operations: Arguments to be const&

carlopi committed:
Interval >= is NOT > or =, but derived from > rules

This requres a minor change in a test, following the same behaviour of postgres

carlopi committed:
DistinctFrom::Operation: Refactor (NFC)

carlopi committed:
normalize Distinct operation to DistinctGreaterThan

Transformation is that first we check if either is null + only in the else branch actually calling GreaterThan.
This allows to separate concerns, allowing to swap either left<->right or left_null<->right_null.

carlopi committed:
normalization: normalize some vector operations

carlopi committed:
sort: Avoid actually building string_t but perform comparisons directly on const char*

carlopi committed:
normalize execute_comparison, at least in the obvious ways

Mytherin commented on src/common/sort/comparators.cpp:
```diff
@@ -163,17 +163,23 @@ int Comparators::CompareStringAndAdvance(data_ptr_t &left_ptr, data_ptr_t &right
-	return TemplatedCompareVal<string_t>((data_ptr_t)&left_val, (data_ptr_t)&right_val);
+
+	if (memcmp_res != 0)
+		return memcmp_res;
```
nit: our coding style enforces the use of brackets for single-line statements, e.g.:

```cpp
if (memcmp_res != 0) {
 		return memcmp_res;
} 		
```

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -66,6 +66,9 @@ struct string_t {
 	bool IsInlined() const {
 		return GetSize() <= INLINE_LENGTH;
 	}
+	void const *getInlined() {
```
nit: `GetInlined`

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -117,10 +120,79 @@ struct string_t {
+#endif
+			uint64_t A;
+			uint64_t B;
+			memcpy(&A, &a, 8u);
```
Can we use `Load<uint64_t>(&A)` here?

carlopi replied:
Thanks!

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -117,10 +120,79 @@ struct string_t {
+				return false;
+			}
+			// they have the same length and same prefix!
+			memcpy(&A, ((const char *)&a + 8u), 8u);
```
Same here, `Load<uint64_t>`

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -117,10 +120,79 @@ struct string_t {
+			memcpy(&A, &a, 8u);
+			memcpy(&B, &b, 8u);
+			if (A != B) {
+				// Either lenght or prefix are different -> not equal
```
typo: `length` not `lenght`

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -117,10 +120,79 @@ struct string_t {
+		static bool GreaterThan(const string_t &left, const string_t &right) {
+			const uint32_t left_length = left.GetSize();
+			const uint32_t right_length = right.GetSize();
+			const uint32_t min_lenght = std::min<uint32_t>(left_length, right_length);
```
typo: `length`

Mytherin commented on src/include/duckdb/common/types/string_type.hpp:
```diff
@@ -117,10 +120,79 @@ struct string_t {
+				return A > B;
+#endif
+			auto memcmp_res = memcmp(left.GetDataUnsafe(), right.GetDataUnsafe(), min_lenght);
+			auto final_res = (memcmp_res == 0) ? (left_length > right_length) : (memcmp_res > 0);
```
Perhaps we can use predication instead of having a branch here, e.g.:
`memcmp_res > 0 || (memcmp_res == 0 && left_length > right_length)`

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments:

Mytherin commented:
Also, it looks like TPC-H Q1 (which uses string hashing and comparisons very intensely) got significantly slower:

```
benchmark/tpch/sf1/q01.benchmark
Old timing: 0.180383
New timing: 0.227067
```

Perhaps have a look there and see if this new comparison/hash actually improves things?

carlopi committed:
Normalize comparisions: Coding style

carlopi committed:
string_type: memcpy -> Load<>

carlopi committed:
string_type: memcpy -> Load<> + fix mistake (left -> right)

carlopi committed:
string_type: Fix spelling

carlopi committed:
string_type: Rework select

carlopi committed:
string_t::GreaterThan: Fix mistake in the algorithm (bswap!) and define

carlopi committed:
Store<> to get const& as parameter

carlopi committed:
string_type: Fix inequality boundaries, <=12 is inlined, >12 is not

carlopi commented:
> Also, it looks like TPC-H Q1 (which uses string hashing and comparisons very intensely) got significantly slower:
> ...
> ```
> benchmark/tpch/sf1/q01.benchmark
> Old timing: 0.180383
> New timing: 0.227067
> ```
> 
> Perhaps have a look there and see if this new comparison/hash actually improves things?

Actually I added back an optimisation that I though was not needed, and now numbers look like this:

CI (there is lots of noise, but seems to be faster)
```
benchmark/tpch/sf1/q01.benchmark
Old timing: 0.16775
New timing: 0.164793

benchmark/tpch/sf1/q19.benchmark
Old timing: 0.145454
New timing: 0.129259
```

MacOS M1 (in all those cases this PR is faster that base branch by at least a few sigma, plus a bunch of tests with unchanged results)
```
micro/cast/cast_date_string.benchmark         a = 0.025s      b = 0.036s      ( 0.010917s     , 30.52%        , 58.020 sigma)
micro/cast/cast_int_string.benchmark          a = 0.024s      b = 0.032s      ( 0.007647s     , 24.23%        , 21.385 sigma)
micro/cast/cast_double_string.benchmark       a = 0.106s      b = 0.118s      ( 0.012924s     , 10.91%        , 33.491 sigma)
micro/cast/cast_timestamp_string.benchmark    a = 0.042s      b = 0.058s      ( 0.016011s     , 27.61%        , 77.429 sigma)
tpch/sf1/q01.benchmark                        a = 0.039s      b = 0.039s      ( 0.000774s     , 1.96%         , 2.494 sigma)
tpch/sf1/q19.benchmark                        a = 0.030s      b = 0.033s      ( 0.003012s     , 9.22%         , 4.262 sigma)
tpcds/sf1/q02.benchmark                       a = 0.028s      b = 0.030s      ( 0.001700s     , 5.66%         , 3.575 sigma)
tpcds/sf1/q04.benchmark                       a = 0.135s      b = 0.146s      ( 0.010830s     , 7.42%         , 6.499 sigma)
tpcds/sf1/q11.benchmark                       a = 0.110s      b = 0.121s      ( 0.011397s     , 9.41%         , 7.115 sigma)
tpcds/sf1/q22.benchmark                       a = 0.132s      b = 0.137s      ( 0.004544s     , 3.32%         , 4.712 sigma)
tpcds/sf1/q36.benchmark                       a = 0.035s      b = 0.039s      ( 0.003493s     , 8.97%         , 4.016 sigma)
tpcds/sf1/q59.benchmark                       a = 0.049s      b = 0.052s      ( 0.002452s     , 4.72%         , 3.571 sigma)
tpcds/sf1/q74.benchmark                       a = 0.063s      b = 0.069s      ( 0.006030s     , 8.72%         , 4.055 sigma)
```

I think this is worth putting in, since fast-path for string_type operations, expecially if inlined, seems to be significant enough.

There are 2 unfinished parts:
* integrate A/B testing script with current regression testing script
This I will do, separating A/B script to a separate PR.

* there is any decent way around *reinterpret_cast<uint32_t>(some_pointer) ?
Load<uint32_t> do not currently propagates alignment informations, and the compilers seems not to be able to recreate them in all cases. On platforms where the penalty for unaligned loads is negligible, this is not a problem, and performances are the same between `Load<>` and `*reinterpret_cast<>`, but in architectures where this is a hard failure, the fact that no alignment information is not propagated is a bit bump (say 30% in tight loops). Here we are leveraging the fact that string_type will always be allocated at least aligned at 4 bytes, so it's members are always aligned to 4 bytes.

(There have been also some unconnected failures in CI, it passed but TPCDS on my fork)

carlopi committed:
Fix test dependency on hashing function

Now test/sql/window/test_window_range should be independent to changes of hashing logic

Mytherin commented:
Thanks!
